lib/puppetserver/ca/puppet_config.rb

```diff
@@ -84,60 +84,68 @@ def resolve_settings(overrides = {})
         substitutions = Hash.new {|h, k| k }
         settings = {}
 
-        confdir = user_specific_conf_dir
-        settings[:confdir] = substitutions['$confdir'] = confdir
-
-        ssldir = overrides.fetch(:ssldir, '$confdir/ssl')
-        settings[:ssldir] = substitutions['$ssldir'] = ssldir.sub('$confdir', confdir)
-
-        certdir = overrides.fetch(:certdir, '$ssldir/certs')
-        settings[:certdir] = substitutions['$certdir'] = certdir.sub(unresolved_setting, substitutions)
-
-        cadir = overrides.fetch(:cadir, '$ssldir/ca')
-        settings[:cadir] = substitutions['$cadir'] = cadir.sub(unresolved_setting, substitutions)
-
-        settings[:certname] = substitutions['$certname'] = overrides.fetch(:certname, @certname)
-
-        server = overrides.fetch(:server, '$certname')
-        settings[:server] = substitutions['$server'] = server.sub(unresolved_setting, substitutions)
-
-        settings[:masterport] = substitutions['$masterport'] = overrides.fetch(:masterport, '8140')
-
-        settings[:ca_name] =  overrides.fetch(:ca_name, 'Puppet CA: $certname')
-        settings[:root_ca_name] = overrides.fetch(:root_ca_name, "Puppet Root CA: #{SecureRandom.hex(7)}")
+        defaults = {
+          # Order for base settings here matters!
+          # These need to be evaluated before we can construct their dependent
+          # defaults below
+          :base => {
+            :confdir => user_specific_conf_dir,
+            :ssldir =>'$confdir/ssl',
+            :cadir =>'$ssldir/ca',
+            :certdir =>'$ssldir/certs',
+            :certname => @certname,
```

Comment:
While you're cleaning up this class can we memoize the value computed in `user_specific_conf*` and wrap `@certname` in a similar reader method?